tesseract_core/runtime/finite_differences.py

```diff
@@ -0,0 +1,551 @@
+    output_patterns = _annotation_to_paths(ad_outputs)
+
+    ad_inputs = []
+    for pattern in input_patterns:
+        ad_inputs.extend(expand_regex_path(pattern, inputs))
+
+    ad_outputs = []
+    for pattern in output_patterns:
+        ad_outputs.extend(expand_regex_path(pattern, outputs))
+
+    return ad_inputs, ad_outputs
+
+
+def _cached_jacobian(fn: Callable) -> Callable:
+    """Cache the result of the jacobian computation based on input_path, output_path, and arr_idx."""
+    cache = {}
+
+    @wraps(fn)
+    def _wrapper(*args: Any, **kwargs: Any) -> Any:
+        _, _, input_path, output_path, arr_idx, *_ = args
+        key = (input_path, output_path, tuple(arr_idx))
+        if key not in cache:
+            try:
+                cache[key] = fn(*args, **kwargs)
+            except Exception as e:
+                cache[key] = e
+        if isinstance(cache[key], Exception):
+            raise cache[key]
+        return cache[key]
+
+    _wrapper.clear_cache = cache.clear
+    return _wrapper
+
+
+@_cached_jacobian
+def _jacobian_via_apply(
+    endpoints_func: dict[str, Callable],
+    inputs: dict[str, Any],
+    input_path: Sequence[str],
+    output_path: Sequence[str],
+    arr_idx: tuple[int, ...],
+    eps: float = 1e-4,
+) -> np.ndarray:
+    """Compute a Jacobian row using central finite differences."""
+    apply_fn = endpoints_func["apply"]
+    ApplySchema = get_input_schema(apply_fn)
+
+    def _perturbed_apply(inputs, eps):
+        input_val = get_at_path(inputs, input_path).copy()
+        if arr_idx:
+            # array
+            input_val[arr_idx] += eps
+        else:
+            # scalar
+            input_val += eps
+        inputs_plus = set_at_path(inputs, {input_path: input_val})
+        return apply_fn(
+            ApplySchema.model_validate({"inputs": inputs_plus})
+        ).model_dump()
+
+    output_plus = _perturbed_apply(inputs, eps)
+    output_minus = _perturbed_apply(inputs, -eps)
+
+    central_diff = (
+        get_at_path(output_plus, output_path) - get_at_path(output_minus, output_path)
+    ) / (2 * eps)
+    return central_diff
+
+
+@_cached_jacobian
+def _jacobian_via_jacobian(
+    endpoints_func: dict[str, Callable],
+    inputs: dict[str, Any],
+    input_path: Sequence[str],
+    output_path: Sequence[str],
+    arr_idx: tuple[int, ...],
+) -> np.ndarray:
+    """Compute a Jacobian row using the jacobian endpoint."""
+    jac_fn = endpoints_func["jacobian"]
+
+    def _jacobian(inputs):
+        JacSchema = get_input_schema(jac_fn)
+        return jac_fn(
+            JacSchema.model_validate(
+                {
+                    "inputs": inputs,
+                    "jac_inputs": [input_path],
+                    "jac_outputs": [output_path],
+                }
+            )
+        ).model_dump()
+
+    output = _jacobian(inputs)
+    output_val = output[output_path][input_path]
+    jac_slice = (slice(None),) * (output_val.ndim - len(arr_idx)) + tuple(arr_idx)
```

Comment:
This is equivalent to `[:, :, ...out_dim-len(arr_idx) times, *arr_idx]`, right? I still have to digest the `-len(arr_idx)` part; wouldn't the jacobian have output_val.ndim + input_val.ndim axes? and arr_idx input_val has len(arr_idx) axes? 🤔 
from what is written here instead, it seems like jacobian has as many axes as output_val.ndim

It seems to work, but maybe there should be a short comment explaining it

Reply:
`output_val` *is* the output from `jacobian`, so that seems like a tautology :)

Added a comment.